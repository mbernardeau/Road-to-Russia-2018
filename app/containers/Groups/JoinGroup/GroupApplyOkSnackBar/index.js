import GroupApplyOkSnackBar from './GroupApplyOkSnackBar';

export default GroupApplyOkSnackBar;
