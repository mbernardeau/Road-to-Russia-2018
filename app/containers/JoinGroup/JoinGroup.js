import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import AutoComplete from 'material-ui/AutoComplete';
import RaisedButton from 'material-ui/RaisedButton';

const dataSourceConfig = {
  text: 'name',
  value: 'id',
};

const createDataSource = (groups) => _.map(groups, (value, key) => ({
  [dataSourceConfig.text]: value.name,
  [dataSourceConfig.value]: key,
}));


class JoinGroup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectionValid: false,
    };
  }

  handleNewRequest = (chosenRequest, index) => {
    if (index !== -1) {
      this.setState({
        selectionValid: true,
        selected: chosenRequest,
      });
    } else if (this.state.selectionValid) {
      this.setState({
        selectionValid: false,
      });
    }
  }

  applyInGroup = () => {
    if (this.state.selectionValid) {
      this.props.applyInGroup(this.props.uid, this.state.selected);
    }
  }

  render() {
    const groups = createDataSource(this.props.groups);

    return (
      <div style={styles.container}>
        <h2>Rejoindre un groupe</h2>
        <div style={styles.fields}>
          <AutoComplete
            floatingLabelText="Nom du groupe"
            dataSource={groups}
            filter={AutoComplete.fuzzyFilter}
            dataSourceConfig={dataSourceConfig}
            onNewRequest={this.handleNewRequest}
          />
          <RaisedButton disabled={!this.state.selectionValid} onClick={this.applyInGroup} label="Envoyer la demande" primary />
        </div>
      </div>
    );
  }
}

const styles = {
  container: {
    textAlign: 'center',
  },

  fields: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
};

JoinGroup.propTypes = {
  groups: PropTypes.objectOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
    })
  ),
  applyInGroup: PropTypes.func.isRequired,
  uid: PropTypes.string.isRequired,
};

export default JoinGroup;
