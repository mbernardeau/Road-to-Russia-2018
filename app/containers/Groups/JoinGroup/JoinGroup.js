import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { map } from 'lodash';

import Button from 'material-ui/Button';
import Card, { CardActions, CardContent } from 'material-ui/Card';
import { MenuItem } from 'material-ui/Menu';
import Select from 'material-ui/Select';
import { FormControl, FormHelperText } from 'material-ui/Form';
import Typography from 'material-ui/Typography';

import GroupApplyOkSnackBar from './GroupApplyOkSnackBar';

class JoinGroup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: '',
    };
  }

  handleSelection = (event) => {
    this.setState({
      selected: event.target.value,
    });
  }

  handleRequestClose = () => {
    this.setState({
      sent: false,
    });
  }

  applyInGroup = () => {
    if (this.state.selected) {
      this.props.applyInGroup(this.props.uid, this.state.selected);
      this.setState({ selected: '', sent: true });
    }
  }

  render() {
    const { groups, disabled } = this.props;

    return (
      <Card style={styles.fields}>

        <Typography gutterBottom type="headline">Rejoindre un groupe</Typography>
        <br />
        <Typography gutterBottom type="subheading">
          Rejoignez un groupe pour vous confrontez à vos amis, collègues, familles...
        </Typography>

        <CardContent style={styles.content}>
          <FormControl disabled={disabled} fullWidth>
            <Select
              value={this.state.selected}
              onChange={this.handleSelection}
              fullWidth
            >
              {map(groups, (value, key) =>
                <MenuItem key={key} value={key}>{value.name}</MenuItem>
              )}
            </Select>
          </FormControl>
          <FormHelperText>
            {disabled ?
              'Aucun groupe à rejoindre' :
              'Rechercher un groupe'
            }
          </FormHelperText>
        </CardContent>

        <CardActions>
          <Button
            style={styles.button}
            disabled={!this.state.selected}
            onClick={this.applyInGroup}
            color="primary"
            raised
          >
            Envoyer la demande
          </Button>
        </CardActions>

        <GroupApplyOkSnackBar
          open={this.state.sent}
          handleRequestClose={this.handleRequestClose}
        />

      </Card>
    );
  }
}

const styles = {
  fields: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    width: '90%',
    marginBottom: 15,
    padding: 15,
  },

  content: {
    width: '90%',
  },
};

JoinGroup.propTypes = {
  groups: PropTypes.objectOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
    })
  ),
  disabled: PropTypes.bool,
  applyInGroup: PropTypes.func.isRequired,
  uid: PropTypes.string.isRequired,
};

export default JoinGroup;
