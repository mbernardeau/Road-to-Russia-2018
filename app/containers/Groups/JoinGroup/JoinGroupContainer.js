import {
  firebaseConnect,
} from 'react-redux-firebase';
import { connect } from 'react-redux';
import { compose } from 'redux';

import JoinGroup from './JoinGroup';

export default compose(
  firebaseConnect([
    { path: 'groups' },
  ]),
  connect(
    ({ firebase: { data: { groups }, auth: { uid } } }) => ({
      groups,
      uid,
    }),
    (dispatch, { firebase }) => ({
      applyInGroup: (uid, groupId) => firebase.set(`users/${uid}/groups/${groupId}`, true),
    }),
  ),
)(JoinGroup);
