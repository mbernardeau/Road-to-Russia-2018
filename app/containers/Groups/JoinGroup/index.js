import JoinGroup from './JoinGroupContainer';

export default JoinGroup;
