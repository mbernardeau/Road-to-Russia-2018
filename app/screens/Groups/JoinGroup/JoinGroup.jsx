import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Button from 'material-ui/Button'
import Card from 'material-ui/Card'
import CardActions from 'material-ui/Card/CardActions'
import CardContent from 'material-ui/Card/CardContent'
import TextField from 'material-ui/TextField'
import FormControl from 'material-ui/Form/FormControl'
import Typography from 'material-ui/Typography'

import GroupApplyStatus from './GroupApplyStatus'

import './JoinGroup.scss'

class JoinGroup extends Component {
  constructor(props) {
    super(props)
    this.state = {
      code: '',
    }
  }

  handleSelection = event => {
    this.setState({
      code: event.target.value,
    })
  }

  applyInGroup = () => {
    if (this.state.code) {
      this.props.applyInGroup(this.state.code)
      this.setState({ code: '' })
    }
  }

  render() {
    const { code } = this.state

    return (
      <Card className="join-group-card">
        <Typography gutterBottom variant="headline">
          Rejoindre une tribu
        </Typography>
        <br />
        <Typography gutterBottom variant="subheading">
          Rejoignez une tribu pour vous confrontez à vos amis, collègues, familles...
        </Typography>

        <CardContent>
          <Typography gutterBottom variant="caption">
            Entrez le code qui vous a été fourni par l&apos;administrateur de la tribu
          </Typography>
          <FormControl>
            <TextField
              required
              id="join-group-code"
              label="Code"
              value={code}
              margin="normal"
              onChange={this.handleSelection}
            />
          </FormControl>
        </CardContent>

        <CardActions>
          <Button
            disabled={!this.state.code}
            onClick={this.applyInGroup}
            color="primary"
            variant="raised"
          >
            Envoyer la demande
          </Button>
        </CardActions>

        <GroupApplyStatus />
      </Card>
    )
  }
}

JoinGroup.propTypes = {
  applyInGroup: PropTypes.func.isRequired,
}

export default JoinGroup
