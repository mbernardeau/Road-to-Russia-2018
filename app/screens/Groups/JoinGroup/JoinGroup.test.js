import React from 'react'
import renderer from 'react-test-renderer'

import JoinGroup from './JoinGroup'

jest.mock('material-ui/Button', () => 'Button')
jest.mock('material-ui/Card', () => 'Card')
jest.mock('material-ui/Card/CardActions', () => 'CardActions')
jest.mock('material-ui/Card/CardContent', () => 'CardContent')
jest.mock('material-ui/Menu/MenuItem', () => 'MenuItem')
jest.mock('material-ui/Select', () => 'Select')
jest.mock('material-ui/Form', () => ({
  FormControl: 'FormControl',
  FormHelperText: 'FormHelperText',
}))
jest.mock('material-ui/Typography', () => 'Typography')

describe('screens/JoinGroup', () => {
  const props = {
    groups: {
      AA: {
        name: 'group 1',
      },
      BB: {
        name: 'group 2',
      },
      CC: {
        name: 'group 3',
      },
    },
    applyInGroup: () => {},
    uid: 'userId',
  }

  it('renders correctly', () => {
    const tree = renderer.create(<JoinGroup {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when disabled', () => {
    const tree = renderer.create(<JoinGroup {...props} disabled />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
