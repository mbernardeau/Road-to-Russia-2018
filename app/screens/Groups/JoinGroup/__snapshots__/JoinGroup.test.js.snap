// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens/JoinGroup renders correctly 1`] = `
<Card
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": 15,
      "padding": 15,
      "width": "90%",
    }
  }
>
  <Typography
    gutterBottom={true}
    type="headline"
  >
    Rejoindre un groupe
  </Typography>
  <br />
  <Typography
    gutterBottom={true}
    type="subheading"
  >
    Rejoignez un groupe pour vous confrontez à vos amis, collègues, familles...
  </Typography>
  <CardContent
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <FormControl
      disabled={undefined}
      fullWidth={true}
    >
      <Select
        fullWidth={true}
        onChange={[Function]}
        value=""
      >
        <MenuItem
          value="AA"
        >
          group 1
        </MenuItem>
        <MenuItem
          value="BB"
        >
          group 2
        </MenuItem>
        <MenuItem
          value="CC"
        >
          group 3
        </MenuItem>
      </Select>
    </FormControl>
    <FormHelperText>
      Rechercher un groupe
    </FormHelperText>
  </CardContent>
  <CardActions>
    <Button
      color="primary"
      disabled={true}
      onClick={[Function]}
      raised={true}
      style={undefined}
    >
      Envoyer la demande
    </Button>
  </CardActions>
</Card>
`;

exports[`screens/JoinGroup renders correctly when disabled 1`] = `
<Card
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": 15,
      "padding": 15,
      "width": "90%",
    }
  }
>
  <Typography
    gutterBottom={true}
    type="headline"
  >
    Rejoindre un groupe
  </Typography>
  <br />
  <Typography
    gutterBottom={true}
    type="subheading"
  >
    Rejoignez un groupe pour vous confrontez à vos amis, collègues, familles...
  </Typography>
  <CardContent
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <FormControl
      disabled={true}
      fullWidth={true}
    >
      <Select
        fullWidth={true}
        onChange={[Function]}
        value=""
      >
        <MenuItem
          value="AA"
        >
          group 1
        </MenuItem>
        <MenuItem
          value="BB"
        >
          group 2
        </MenuItem>
        <MenuItem
          value="CC"
        >
          group 3
        </MenuItem>
      </Select>
    </FormControl>
    <FormHelperText>
      Aucun groupe à rejoindre
    </FormHelperText>
  </CardContent>
  <CardActions>
    <Button
      color="primary"
      disabled={true}
      onClick={[Function]}
      raised={true}
      style={undefined}
    >
      Envoyer la demande
    </Button>
  </CardActions>
</Card>
`;
